File: app/src/main/java/com/example/myapplication/adapters/MemeAdapter.java
package com.example.myapplication.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;
import com.example.myapplication.R;
import com.example.myapplication.models.MemeDto;
import com.squareup.picasso.Picasso;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class MemeAdapter extends RecyclerView.Adapter<MemeAdapter.MemeHolder> {

    private List<MemeDto> memes = new ArrayList<>();


    @NonNull
    @Override
    public MemeHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.fragment_meme_block, parent, false);
        return new MemeHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MemeHolder holder, int position) {
        holder.bind(memes.get(position));
    }

    @Override
    public int getItemCount() {
        return memes.size();
    }

    public void setItems(Collection<MemeDto> memes) {
        this.memes.addAll(memes);
        notifyDataSetChanged();
    }

    public void clearItems() {
        memes.clear();
        notifyDataSetChanged();
    }

    public class MemeHolder extends RecyclerView.ViewHolder {
        private TextView title;
        private ImageView image;

        public MemeHolder(@NonNull View itemView) {
            super(itemView);
            title=itemView.findViewById(R.id.titleMeme);
            image=itemView.findViewById(R.id.imageMeme);
        }

        public void bind(MemeDto meme){
            title.setText(meme.getTitle());
            Picasso.get().load(meme.getPhotoUtl()).into(image);
        }
    }
}

File: app/src/main/java/com/example/myapplication/adapters/MemeAdapter.kt
package com.example.myapplication.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.R
import com.example.myapplication.models.MemeDto
import com.squareup.picasso.Picasso
import java.text.SimpleDateFormat
import java.util.Date
import java.util.*

class MemeAdapter : RecyclerView.Adapter<MemeAdapter.MemeHolder>() {

    private val memes = ArrayList<MemeDto>()
    private val mListeners: MutableList<ChangeListener> = mutableListOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MemeHolder {
        val view = LayoutInflater.from(parent.context).inflate(
            com.example.myapplication.R.layout.fragment_meme_block,
            parent,
            false
        )
        return MemeHolder(view).listen { position, _ ->
            changeSource(position)
        }
    }

    override fun onBindViewHolder(holder: MemeHolder, position: Int) {
        val item = memes[position]
        holder.bind(item)
    }

    override fun getItemCount(): Int {
        return memes.size
    }

    fun setItems(memes: Collection<MemeDto>) {
        this.memes.addAll(memes)
        notifyDataSetChanged()
    }

    fun clearItems() {
        memes.clear()
        notifyDataSetChanged()
    }

    inner class MemeHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val titlePrev: TextView = itemView.findViewById(R.id.titleMeme)
        private val imagePrev: ImageView = itemView.findViewById(R.id.imageMeme)

        fun bind(meme: MemeDto) {
            titlePrev.text = meme.title
            Picasso.get().load(meme.photoUtl).into(imagePrev)
        }
    }

    fun addListener(listener: ChangeListener){
        mListeners.add(listener)
    }

    fun changeSource(position: Int){
        mListeners.forEach{
            it.sourceChanged(position)
        }
    }

    fun <T: RecyclerView.ViewHolder> T.listen(event: (position: Int,type:Int) -> Unit):T{
        itemView.setOnClickListener {
            event.invoke(adapterPosition, itemViewType)
        }
        return this
    }

    operator fun get(position: Int):MemeDto{
        return memes[position]
    }

    fun getFormattedDate(rawDate: String) : String {
        return SimpleDateFormat("dd/MM/yyyy HH:mm:ss").format(Date((rawDate.toLong())*1000))
    }
}
